README.md
ctxerr
======

[![MIT license](https://img.shields.io/badge/license-MIT-blue.svg)](LICENSE)
[![GoDoc](https://godoc.org/github.com/nochso/ctxerr?status.svg)](https://godoc.org/github.com/nochso/ctxerr)
[![GitHub release](https://img.shields.io/github/release/nochso/ctxerr.svg)](https://github.com/nochso/ctxerr/releases)
[![Go Report Card](https://goreportcard.com/badge/github.com/nochso/ctxerr)](https://goreportcard.com/report/github.com/nochso/ctxerr)
[![Build Status](https://travis-ci.org/nochso/ctxerr.svg?branch=master)](https://travis-ci.org/nochso/ctxerr)
[![Coverage Status](https://coveralls.io/repos/github/nochso/ctxerr/badge.svg?branch=master)](https://coveralls.io/github/nochso/ctxerr?branch=master)

**ctxerr** is a [Go][] library and CLI utility for pretty-printing linter/parser errors.

Instead of just describing what's wrong, it lets you point at the wrong input:

```
narf.txt:1:8: string is missing closing quote
 1 | "foo bar
   |         ^ missing closing quote
```

- Create pretty error messages for linters, parsers, etc. by pointing at the
  relevant position in the source code.
- Parse existing error messages into `Ctx` structs.
- CLI utility `ctx` scans STDIN for linter errors and enhances them with the
  reported source line.

---

- [Installation](#installation)
- [Documentation](#documentation)
- [Changes](#changes)
- [License](#license)


Installation
------------


### Library

```bash
go get -u github.com/nochso/ctxerr
```

### CLI utility `ctx`

```bash
go install github.com/nochso/ctxerr/cmd/ctx
```


Documentation
-------------

### Library

The [GoDoc](https://godoc.org/github.com/nochso/ctxerr) pages contain plenty
of examples.


### CLI utility `ctx`

```shell
$ ctx -h
ctx 1.0.0-beta
Pretty prints parser errors from stdin.

Possible input:
  path/file.ext:1:5: some error on line 1, column 5
  file.ext:123: column is optional and so is the message:
  file.ext:1

Usage:
  ctx < log.txt
  gometalinter . | ctx

Flags:
  -context NUM
        print NUM lines of context surrounding an error. negative for all, positive for limited and 0 for none (default)
  -no-color
        disable any color output
  -pessimistic
        print only matching errors, ignore everything else
```


Changes
-------

All notable changes to this project will be documented in the [changelog].

The format is based on [Keep a Changelog](http://keepachangelog.com/) and this
project adheres to [Semantic Versioning](http://semver.org/).


License
-------

This project is released under the [MIT license](LICENSE).


[changelog]: CHANGELOG.md
[releases]: https://github.com/nochso/ctxerr/releases
[Go]: https://golang.org

CHANGELOG.md
Change Log
==========

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/) and this
project adheres to [Semantic Versioning](http://semver.org/).


[Unreleased]
============

### Added

- Bool `NoColor` to disable color output separately from fatih/color.
- Method `Ctx.ErrorLine() string` returns a single line error message.
- Parse error messages printed by `go test`.
- Command line arguments are run as given. Stdout and stderr are pretty printed
  along the way.


[1.0.0] - 2017-01-21
====================

### Added

- Global integer `DefaultContext` allows you to override the default amount of
  context lines (default 0).
- Function `NewFromPath(path string, region ctxerr.Region) (ctxerr.Ctx, error)`
  returns a new Ctx based on a path to an existing file.
- Method `Ctx.Error()` implementing the error interface, replacing `Ctx.String()`.
- Function `Parse(line string) (*Ctx, error)` parses typical linter output into
  Ctx structs.
  Returns error `ErrNoMatch` when unlikely to be a linter error.
  - Currently supports output of `gometalinter` and `npm run lint`.
- Command `ctx` enhances stdin with pretty errors and optional context.

### Changed

- Exported Position (a Region consists of two position).
- Exported all fields of Ctx and Region.
- Proper handling of Region with zero columns. It is now treated as a full line.

### Removed

- Methods of `Ctx`: `WithPath`, `WithContext` and `WithHint` as they're now exported as fields.
- Struct `Error` and method `Ctx.ToError()`. Use the exported `Err` field of `Ctx` instead.

### Fixed

- Tabs are now properly handled when padding and pointing to regions.
- No more panic when pointing outside of the line (to the right).


0.1.0
=====

### Added

- Initial public release.


[Unreleased]: https://github.com/nochso/ctxerr/compare/1.0.0...HEAD
[1.0.0]: https://github.com/nochso/ctxerr/compare/0.1.0...1.0.0

ctx_test.go
package ctxerr

import (
	"fmt"
	"os"
	"testing"
)

func init() {
	NoColor = true
}

func ExampleNew() {
	fmt.Println(New("source code containing error", Range(1, 24, 1, 28)))
	// Output:
	// 1:24-28:
	// 1 | source code containing error
	//   |                        ^^^^^
}

func ExamplePoint() {
	fmt.Println(New("00100", Point(1, 3)))
	// Output:
	// 1:3:
	// 1 | 00100
	//   |   ^
}

func ExampleRange() {
	fmt.Println(New("01110", Range(1, 2, 1, 4)))
	// Output:
	// 1:2-4:
	// 1 | 01110
	//   |  ^^^
}

func ExampleRange_multiline() {
	fmt.Println(New("00001\n11110", Range(1, 5, 2, 4)))
	// Output:
	// 1:5-2:4:
	// 1 | 00001
	//   |     ^
	// 2 | 11110
	//   | ^^^^
}

func ExampleCtx_Error_hint() {
	ctx := New("010101102110", Point(1, 9))
	ctx.Hint = "don't worry, bender"
	fmt.Println(ctx)
	// Output:
	// 1:9:
	// 1 | 010101102110
	//   |         ^ don't worry, bender
}

func ExampleCtx_Error_contextLimited() {
	in := `1st
2nd
3rd has an error
4th
5th`
	ctx := New(in, Point(3, 12))
	ctx.Context = 1
	fmt.Println(ctx)
	// Output:
	// 3:12:
	// 2 | 2nd
	// 3 | 3rd has an error
	//   |            ^
	// 4 | 4th
}

func ExampleCtx_Error_contextLimitedMultiline() {
	in := `1st
2nd
3rd has an error
4th still has an error
5th`
	ctx := New(in, Range(3, 1, 4, 22))
	ctx.Context = 1
	fmt.Println(ctx)
	// Output:
	// 3:1-4:22:
	// 2 | 2nd
	// 3 | 3rd has an error
	//   | ^^^^^^^^^^^^^^^^
	// 4 | 4th still has an error
	//   | ^^^^^^^^^^^^^^^^^^^^^^
	// 5 | 5th
}

func ExampleCtx_Error_contextAll() {
	in := `1st
2nd
3rd has an error
4th
5th`
	ctx := New(in, Point(3, 12))
	ctx.Context = -1
	fmt.Println(ctx)
	// Output:
	// 3:12:
	// 1 | 1st
	// 2 | 2nd
	// 3 | 3rd has an error
	//   |            ^
	// 4 | 4th
	// 5 | 5th
}

func ExampleCtx_Error_path() {
	ctx := New("42", Point(1, 1))
	ctx.Path = "/tmp/ctxerr/answer.txt"
	fmt.Println(ctx)
	// Output:
	// /tmp/ctxerr/answer.txt:1:1:
	// 1 | 42
	//   | ^
}

func ExampleCtx_Error() {
	err := New("ab!cd", Point(1, 3))
	err.Err = fmt.Errorf("not a letter")
	fmt.Println(err)
	// Output:
	// not a letter
	// 1:3:
	// 1 | ab!cd
	//   |   ^
}

func ExampleCtx_Error_tabwidth() {
	in := "\tfoo\tbar"
	ctx := New(in, Point(1, 5))
	fmt.Println(ctx)
	// Output:
	// 1:5:
	// 1 |     foo    bar
	//   |        ^^^^
}

func ExampleNewFromPath() {
	cerr, err := NewFromPath("LICENSE", Range(1, 1, 1, 3))
	if err != nil {
		fmt.Println(err)
		return
	}
	cerr.Context = 0
	fmt.Println(cerr)
	// Output:
	// LICENSE:1:1-3:
	// 1 | MIT License
	//   | ^^^
}

func ExampleCtx_ErrorLine() {
	ctx := New("foo", Point(1, 1))
	ctx.Err = fmt.Errorf("something went wrong")
	ctx.Path = "bar.txt"
	fmt.Println(ctx.ErrorLine())
	// Output:
	// bar.txt:1:1: something went wrong
}

func TestNewFromPath_NoSuchFile(t *testing.T) {
	_, err := NewFromPath("foo.txt", Point(1, 1))
	if err == nil {
		t.Error("expected error, got nil")
	}
	if !os.IsNotExist(err) {
		t.Errorf("expected %#v, got %#v", os.ErrNotExist.Error(), err.Error())
	}
}

func testCmp(t *testing.T, exp, act string) {
	if exp != act {
		t.Errorf("expected %#v, got %#v", exp, act)
	}
}

func TestCtx_Error_contextBoundaries(t *testing.T) {
	ctx := New("x", Point(1, 1))
	ctx.Context = 42
	exp := `1:1:
1 | x
  | ^
`
	testCmp(t, exp, ctx.Error())
}

func TestCtx_Error_pointsOutside(t *testing.T) {
	ctx := New("'who needs closing quotes?", Point(1, 27))
	exp := `1:27:
1 | 'who needs closing quotes?
  |                           ^
`
	testCmp(t, exp, ctx.Error())
}

func TestCtx_Error_nilColumn(t *testing.T) {
	ctx := New("foo", Point(1, 0))
	exp := `1:
1 | foo
  | ^^^
`
	testCmp(t, exp, ctx.Error())
}


err_test.go
package ctxerr

import (
	"fmt"
	"testing"
)

func ExampleError_Error() {
	cause := fmt.Errorf("foo must not be negative")
	ctx := New("foo = -1", Point(1, 7))
	err := NewError(cause, ctx)
	fmt.Println(err)
	// Output:
	// foo must not be negative
	// 1:7:
	// 1 | foo = -1
	//   |       ^
}

func TestError_Inner(t *testing.T) {
	cause := fmt.Errorf("foo must not be negative")
	ctx := New("foo = -1", Point(1, 7))
	err := NewError(cause, ctx)
	if err.Inner != cause {
		t.Fatalf("expected %s; got %s", cause, err.Inner)
	}
}


parser_test.go
package ctxerr

import (
	"flag"
	"os"
	"strings"
	"testing"

	"github.com/nochso/golden"
)

var update = flag.Bool("update", false, "update golden test files")

func TestParse(t *testing.T) {
	golden.TestDir(t, "test-fixtures/parse-ok", func(tc golden.Case) {
		ctx, err := Parse(tc.In.String())
		if err != nil {
			t.Errorf("expected nil error; got %#v", err.Error())
			return
		}
		if *update {
			tc.Out.Update([]byte(ctx.Error()))
		}
		tc.Diff(ctx.Error())
	})
}

func TestParse_NoMatch(t *testing.T) {
	input := []string{
		"",
		" ",
		":",
		"a:",
		"a:5",
	}
	for _, in := range input {
		_, err := Parse(in)
		if err != ErrNoMatch {
			t.Errorf("expected %#v, got %#v", ErrNoMatch.Error(), err.Error())
		}
	}
}

func TestParse_NoSuchFile(t *testing.T) {
	_, err := Parse("foo.txt:1:")
	if !os.IsNotExist(err) {
		t.Errorf("expected %#v, got %#v", os.ErrNotExist.Error(), err.Error())
	}
}

func TestParse_InvalidInt(t *testing.T) {
	_, err := Parse("a:9999999999999999999999:")
	if err == nil {
		t.Error("expected error, got nil")
	}
	prefix := "strconv"
	if !strings.HasPrefix(err.Error(), prefix) {
		t.Errorf("expected error to have prefix %#v, got %#v", prefix, err.Error())
	}
}


ctx.go
package ctxerr

import (
	"bufio"
	"bytes"
	"fmt"
	"io/ioutil"
	"strconv"
	"strings"
	"unicode/utf8"

	"github.com/fatih/color"
)

var (
	// DefaultGutter separates line numbers from content.
	DefaultGutter = " | "
	// DefaultPointer is the rune that points up at a region.
	DefaultPointer = '^'
	// DefaultContext is the default amount of context lines surrounding an error.
	// It is used by New* functions.
	DefaultContext = 0
	// NoColor disables color output.
	NoColor bool
)

var (
	red = color.New(color.FgRed)
)

// Ctx points to runes in (multiline) strings.
type Ctx struct {
	Lines []string
	// Context is the maximum amount of context lines surrounding the marked region.
	//
	//	 0: no lines of context.
	//	-1: all lines, the full input string
	//	 3: limited context of 3 lines
	Context int
	Region
	// Path to the source of the context (optional).
	Path string
	// Hint that is displayed near the region markers (optional).
	Hint string
	// Err is the error that occurred at this region (optional).
	Err error
}

// New returns a new Ctx pointing to a region in an input string.
//
// Use functions Point and Range to create a Region.
func New(input string, region Region) Ctx {
	return Ctx{
		Lines:   split(input, region),
		Region:  region,
		Context: DefaultContext,
	}
}

// NewFromPath returns a new Ctx pointing to a region in the given file.
// Returns an error when the file does not exist or could not be read.
func NewFromPath(path string, region Region) (Ctx, error) {
	b, err := ioutil.ReadFile(path)
	if err != nil {
		return Ctx{}, err
	}
	c := New(string(b), region)
	c.Path = path
	return c, nil
}

func split(s string, r Region) []string {
	sc := bufio.NewScanner(strings.NewReader(s))
	l := []string{}
	for sc.Scan() {
		l = append(l, sc.Text())
	}
	return l
}

// Error formats a summary of this context error.
func (c Ctx) Error() string {
	buf := &bytes.Buffer{}
	if c.Err != nil {
		fmt.Fprintf(buf, "%s\n", c.Err)
	}
	if c.Path != "" {
		fmt.Fprintf(buf, "%s:", c.Path)
	}
	fmt.Fprintf(buf, "%s:\n", c.Region)
	start, end := c.lineIndex()
	// length of highest line number
	linePosMaxLen := posLen(end)
	for i, line := range c.Lines[start:end] {
		linePos := start + i + 1
		// write line no. gutter and actual line
		c.writeLineGutter(buf, linePos, linePosMaxLen)
		buf.WriteString(strings.Replace(line, "\t", "    ", -1))
		buf.WriteByte('\n')
		if linePos < c.Start.Line || linePos > c.End.Line {
			// this was just context, don't point at it
			continue
		}
		// this line is being pointed at
		c.writeLineGutter(buf, 0, linePosMaxLen)
		buf.WriteString(strings.Repeat(" ", c.getPad(linePos)))
		buf.WriteString(paint(red, "%s", strings.Repeat(string(DefaultPointer), c.getDots(linePos, line))))
		if c.Hint != "" && c.Start.Line == linePos {
			fmt.Fprintf(buf, " %s", c.Hint)
		}
		buf.WriteString("\n")
	}
	return buf.String()
}

// ErrorLine returns a single line error message.
func (c Ctx) ErrorLine() string {
	s := c.Path
	if s != "" {
		s += ":"
	}
	s += c.Region.String() + ": " + c.Err.Error()
	return s
}

func paint(c *color.Color, format string, a ...interface{}) string {
	if NoColor {
		return fmt.Sprintf(format, a...)
	}
	return c.SprintfFunc()(format, a...)
}

// start and end index of Ctx.Lines including lines of context.
func (c Ctx) lineIndex() (start, end int) {
	if c.Context < 0 {
		return 0, len(c.Lines)
	}
	start = c.Start.Line - c.Context - 1
	if start < 0 {
		start = 0
	}
	end = c.End.Line + c.Context
	if end > len(c.Lines) {
		end = len(c.Lines)
	}
	return
}

func posLen(i int) int {
	return len(strconv.Itoa(i))
}

func (c Ctx) getDots(pos int, line string) int {
	start := 0
	end := len(line)
	if c.Start.Line == pos && c.Start.Col != 0 {
		start = c.Start.Col - 1
	}
	if c.End.Line == pos && c.End.Col != 0 {
		end = c.End.Col
	}

	// How far do we go "over the line"?
	// Remember the extra and limit the end to avoid index errors.
	extra := 0
	if end > len(line) {
		extra = end - len(line)
		end = len(line)
	}
	s := line[start:end]
	// add extra dots over the line
	return extra + utf8.RuneCountInString(s) + strings.Count(s, "\t")*3
}

func (c Ctx) getPad(pos int) int {
	if c.isMultiLine() && c.Start.Line != pos {
		return 0
	}
	if (c.Start.Line == pos && c.Start.Col == 0) || (c.End.Line == pos && c.End.Col == 0) {
		return 0
	}
	pad := c.Start.Col - 1
	pad += strings.Count(c.Lines[c.Start.Line-1][:pad], "\t") * 3
	return pad
}

func (c Ctx) writeLineGutter(buf *bytes.Buffer, line, maxLen int) {
	pad := maxLen  // assume 0, meaning no line info
	if line != 0 { // otherwise exclude line no. from padding
		pad -= posLen(line)
	}
	buf.WriteString(strings.Repeat(" ", pad))
	if line != 0 {
		buf.WriteString(strconv.Itoa(line))
	}
	buf.WriteString(DefaultGutter)
}


parser.go
package ctxerr

import (
	"errors"
	"fmt"
	"regexp"
	"strconv"
)

var (
	// ErrNoMatch is returned by Parse when no file path or position can be found.
	ErrNoMatch       = fmt.Errorf("line does not match position pattern: %s", matchRegionColon.String())
	matchRegionColon = regexp.MustCompile(`^\s*(.+?):(\d+):(\d+)?:?\s*(.*)$`)
	matchRegionNpm   = regexp.MustCompile(`^(.+?)\[(\d+), (\d+)\]: (.*)$`)
)

// Parse a single line with at least a file path and position.
// Returns error ErrNoMatch when this does not appear to be an error line.
// Otherwise will return parsing errors or nil on success.
//
//	foo.go:15:1: message
//	foo.go[1, 2]: message
func Parse(line string) (*Ctx, error) {
	ctx, err := parseRegionColon(line)
	if err != ErrNoMatch {
		return ctx, err
	}
	ctx, err = parseRegionNpm(line)
	return ctx, err
}

func parseRegionColon(line string) (*Ctx, error) {
	m := matchRegionColon.FindStringSubmatch(line)
	if len(m) == 0 {
		return nil, ErrNoMatch
	}

	// require a line number, column may be zero
	lineNo, err := strconv.Atoi(m[2])
	col, _ := strconv.Atoi(m[3])
	if err != nil {
		return nil, err
	}
	ctx, err := NewFromPath(m[1], Point(lineNo, col))
	ctx.Err = errors.New(m[4])
	return &ctx, err
}

func parseRegionNpm(line string) (*Ctx, error) {
	m := matchRegionNpm.FindStringSubmatch(line)
	if len(m) == 0 {
		return nil, ErrNoMatch
	}

	// require a line number, column may be zero
	lineNo, err := strconv.Atoi(m[2])
	col, _ := strconv.Atoi(m[3])
	if err != nil {
		return nil, err
	}
	ctx, err := NewFromPath(m[1], Point(lineNo, col))
	ctx.Err = errors.New(m[4])
	return &ctx, err
}


region.go
package ctxerr

import "fmt"

// Position of a single point in a multiline string.
type Position struct {
	Line, Col int
}

// String representation of a Position.
//
//	1
//	1:2
func (p Position) String() string {
	if p.Col == 0 {
		return fmt.Sprintf("%d", p.Line)
	}
	return fmt.Sprintf("%d:%d", p.Line, p.Col)
}

// Region defines a selection of runes (utf8 codepoints) in a string.
type Region struct {
	Start, End Position
}

// Point returns a Region pointing to a specific rune.
// Line and column are one-based.
func Point(line, col int) Region {
	return Range(line, col, line, col)
}

// Range returns a Region pointing to a range of runes.
// Line and column are one-based.
func Range(startLine, startCol, endLine, endCol int) Region {
	return Region{
		Start: Position{startLine, startCol},
		End:   Position{endLine, endCol},
	}
}

// String representation of a region.
//
//	3       // complete 3rd line
//	1-3     // lines 1 through 3
//	1:1     // pointer
//	1:1-2   // range on single line
//	1:1-2:1 // range over multiple lines
func (r Region) String() string {
	if r.isPointer() {
		return r.Start.String()
	}
	if r.isMultiLine() {
		return fmt.Sprintf("%s-%s", r.Start.String(), r.End.String())
	}
	if r.End.Col == 0 {
		return r.End.String()
	}
	return fmt.Sprintf("%s-%d", r.Start.String(), r.End.Col)
}

func (r Region) isPointer() bool {
	return r.Start.Line == r.End.Line && r.Start.Col == r.End.Col && r.Start.Col != 0
}

func (r Region) isMultiLine() bool {
	return r.Start.Line != r.End.Line
}


cmd/ctx/main.go
package main

import (
	"bufio"
	"flag"
	"fmt"
	"io"
	"os"
	"os/exec"
	"sync"

	"github.com/nochso/ctxerr"
)

var (
	// Version as specified by ldflags on build or this default.
	Version = "1.0.0"
	// BuildDate is optional and can be supplied by ldflags.
	BuildDate = ""
)

var (
	context     = flag.Int("context", ctxerr.DefaultContext, "print `NUM` lines of context surrounding an error. negative for all, positive for limited and 0 for none (default)")
	pessimistic = flag.Bool("pessimistic", false, "print only matching errors, ignore everything else")
)

func usage() {
	bd := BuildDate
	if bd != "" {
		bd = fmt.Sprintf(" (built %s)", bd)
	}
	format := `ctx %s%s
Pretty prints parser errors from stdin or any given command.

Possible input:
  path/file.ext:1:5: some error on line 1, column 5
  file.ext:123: column is optional and so is the message:
  file.ext:1
  parser_test.go[1, 15]: Missing semicolon

Usage:
  ctx < log.txt         # Existing file
  gometalinter . | ctx  # Stdin
  ctx go test           # Separate stdin and stderr of given command
  go test 2>&1 | ctx    # Combine stdin and stderr

Flags:
`
	fmt.Printf(format, Version, bd)
	flag.PrintDefaults()
}

func main() {
	flag.BoolVar(&ctxerr.NoColor, "no-color", false, "disable any color output")
	flag.Usage = usage
	flag.Parse()

	if flag.NArg() == 0 {
		err := scan(os.Stdin, os.Stdout)
		if err != nil {
			fmt.Println(err)
		}
		return
	}
	scanCmd()
}

func scan(r io.Reader, w io.Writer) error {
	sc := bufio.NewScanner(r)
	for sc.Scan() {
		cerr, err := ctxerr.Parse(sc.Text())
		if err != nil {
			if err != ctxerr.ErrNoMatch {
				fmt.Fprintln(w, err)
			}
			if !*pessimistic {
				fmt.Fprintln(w, sc.Text())
			}
			continue
		}
		cerr.Context = *context
		fmt.Fprintln(w, cerr)
	}
	return sc.Err()
}

func scanCmd() {
	cmd := exec.Command(flag.Arg(0), flag.Args()[1:]...)
	stdout, err := cmd.StdoutPipe()
	if err != nil {
		panic(err)
	}
	stderr, err := cmd.StderrPipe()
	if err != nil {
		panic(err)
	}
	var wg sync.WaitGroup
	wg.Add(2)
	errs := make(chan error)
	go func() { errs <- scan(stdout, os.Stdout) }()
	go func() { errs <- scan(stderr, os.Stderr) }()
	go func() {
		wg.Wait()
		close(errs)
	}()
	err = cmd.Run()
	if err != nil {
		fmt.Println(err)
	}
	for err := range errs {
		if err != nil {
			fmt.Println(err)
		}
		wg.Done()
	}
}
